### PR TITLE
Add pointer type covariance

### PR DESCRIPTION
**Pull request by AutonomicPerfectionist**
**Is this pull request associated with an issue(s)?**
No.

**Description**

This PR further enriches the type system by allowing pointer types to be checked for covariance. What this means is that a `desc*` can now be used where a `str*` is expected, whereas previously such a conversion would not be possible due to the invariant nature of the implementation. As such, pointer type inference has been enabled, meaning that the type system will be able to infer whether a string is a pointer and recursively dereferences and infers its type. So now, if we have:

```cmake
set(a_ptr TRUE)
cpp_type_of(a_type a_ptr)
```
The value of `${a_type}` will be `bool*` instead of `desc` like it would be previously. Additionally, `bool*` will now be convertible to `str*`

**TODOs**

- [ ] Update documentation
- [ ] Add new tests
- [ ] Fix old tests

### CODE REVIEW

File: tests/types/implicitly_convertible.cmake [hunk 1]
```diff
@@ -318,12 +330,24 @@ function("${test_cpp_implicitly_convertible}")
...
+
     ct_add_section(NAME "double_pointer_to_double_pointer")
     function("${double_pointer_to_double_pointer}")
         cpp_implicitly_convertible(result bool** bool**)
         ct_assert_equal(result TRUE)
     endfunction()
 
+    ct_add_section(NAME "double_pointer_to_desc_pointer")
+    function("${double_pointer_to_desc_pointer}")
+        cpp_implicitly_convertible(result bool** desc*)
```
Comment (ryanmrichard) [inline tests/types/implicitly_convertible.cmake:347]: Nice edge-case catch. I wouldn't thought of this.

**Review by ryanmrichard — APPROVED** [review]

**Review by zachcran — APPROVED** [review]
LGTM. Make sure you have checked that CMaize and CMakeTest tests are still passing as well before merging, please!

**Comment by ryanmrichard** [comment]
Unless I'm missing something, if this breaks CMaize or CMakeTests, it's probably those repose that need to change. Point being, I think we can merge this and deal with the fallout in other PRs.
